### PR TITLE
Download metadata lambda

### PR DESCRIPTION
**Pull request by MancunianSam**
Add the download metadata and files lambda and associated queues, topics etc.
Also added a test and a deploy workflow for GitHub actions. The tests won't pass until https://github.com/nationalarchives/dr2-github-actions/pull/1 is merged.

### CODE REVIEW

File: download-metadata-and-files.tf [hunk 1]
```diff
@@ -0,0 +1,56 @@
+locals {
+  disaster_recovery_bucket_name = "${local.environment}-disaster-recovery"
+  lambda_name                   = "${local.environment}-download-files-metadata"
+  queue_name                    = "${local.environment}-download-files-metadata"
```
Comment (kurtismash) [inline download-metadata-and-files.tf:4]: Should these contain the word `and`?

File: download-metadata-and-files.tf [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  source                   = "git::https://github.com/nationalarchives/da-terraform-modules//sqs"
+  queue_name               = local.queue_name
+  sqs_policy               = templatefile("./templates/sqs/sqs_access_policy.json.tpl", { role_arn = module.download_metadata_and_files_lambda.lambda_role_arn, account_id = var.dp_account_number, queue_name = local.queue_name })
+  redrive_maximum_receives = 3
+  tags = {
+    CreatedBy = "dp-terraform-environments"
+  }
+}
+
+module "disaster_recovery_bucket" {
```
Comment (kurtismash) [inline download-metadata-and-files.tf:48]: Maybe we should break this out? This bucket will also be needed by our DR workflow on-prem - of which they'll be some more resources based on it.

File: templates/deploy_lambda_policy.json.tpl [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+      "Action": [
+        "lambda:GetFunctionConfiguration",
+        "lambda:PublishVersion",
+        "lambda:UpdateFunctionCode",
+        "s3:GetObject",
+        "s3:PutObject"
+      ],
+      "Effect": "Allow",
+      "Resource": [
+        "${download_metadata_lambda_arn}"
```
Comment (kurtismash) [inline templates/deploy_lambda_policy.json.tpl:13]: No0b question - does this need an S3 resource as well?

File: templates/deploy_lambda_policy.json.tpl
Comment (kurtismash) [inline templates/deploy_lambda_policy.json.tpl:1]: Is this file a duplicate?

File: download-metadata-and-files.tf [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  tags = {
+    Name      = local.lambda_name
+    CreatedBy = "dp-terraform-environments"
+  }
+}
+
+module "download_files_sqs" {
+  source                   = "git::https://github.com/nationalarchives/da-terraform-modules//sqs"
+  queue_name               = local.queue_name
+  sqs_policy               = templatefile("./templates/sqs/sqs_access_policy.json.tpl", { role_arn = module.download_metadata_and_files_lambda.lambda_role_arn, account_id = var.dp_account_number, queue_name = local.queue_name })
```
Comment (kurtismash) [inline download-metadata-and-files.tf:41]: Why does this lambda need to send to SQS?

**Review by kurtismash — COMMENTED** [review]
Looks good, a few comments. Also is it worth renaming the `.tf` files for lambdas to make it clear these are lambda functions. An alternative could be grouping lambdas in files based on their workflow? I.e. one file for the event generators, one for the DR workflow, one for the closure workflow, etc.

File: download-metadata-and-files.tf (hunk 1 above)
Comment (MancunianSam) [reply download-metadata-and-files.tf]: I suppose so, I've added it in.

File: download-metadata-and-files.tf (hunk 2 above)
Comment (MancunianSam) [reply download-metadata-and-files.tf]: Yeah, fair point. I've put it into a different file.

File: templates/deploy_lambda_policy.json.tpl (hunk 3 above)
Comment (MancunianSam) [reply templates/deploy_lambda_policy.json.tpl]: Yep, it definitely does. I've added one in.

File: templates/deploy_lambda_policy.json.tpl
Comment (MancunianSam) [reply templates/deploy_lambda_policy.json.tpl:1]: It is, I've deleted it.

File: download-metadata-and-files.tf (hunk 4 above)
Comment (MancunianSam) [reply download-metadata-and-files.tf]: It doesn't need to send to SQS. This is supposed to be the queue that triggers the lambda. I've got it the wrong way round. I've changed it now so it is the queue that triggers the lambda.

File: templates/sqs/sqs_access_policy.json.tpl
Comment (kurtismash) [inline templates/sqs/sqs_access_policy.json.tpl:1]: Is this policy not too permissive? Applying the principal of least privilege here should only permit the SNS topic to `sqs:SendMessage` to the queue.

File: templates/deploy_lambda_policy.json.tpl (hunk 3 above)
Comment (kurtismash) [reply templates/deploy_lambda_policy.json.tpl]: As we're doing object functions I think we need the `/*` on the end of the bucket resource.

**Review by kurtismash — APPROVED** [review]

**Comment by MancunianSam** [comment]
I've also set up terraform to use an external ID from the da-terraform-configurations repo.

File: templates/sqs/sqs_access_policy.json.tpl
Comment (kurtismash) [reply templates/sqs/sqs_access_policy.json.tpl:1]: Is it worth limiting the principal as well?

Comment (MancunianSam) [reply templates/sqs/sqs_access_policy.json.tpl:1]: I've put a TODO in to change it when the SNS topic exists.

**Review by kurtismash — APPROVED** [review]

**Pull request merged**